File: leet_lint_code/lint_code/three_sum.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 57.

#ifndef LEETLINTCODE_LINTCODE_THREESUM_CPP
#define LEETLINTCODE_LINTCODE_THREESUM_CPP

#include <stdio.h>
#include <vector>
#include <algorithm>
#include <map>

class ThreeSum {
 public:
   std::vector<std::vector<int> > threeSum(std::vector<int> num) {
     std::vector<std::vector<int> > ret;
     std::sort(num.begin(),num.end());
     int len = (int)num.size(), j, k;
     for(int i=0; i < len;){
       j = i + 1, k= len - 1;
       while (j < k) {
         int sum = num[i] + num[j] + num[k];
         if(sum > 0){
           while (num[k] == num[k - 1]) { -- k; }
           -- k;
         }else if(sum < 0){
           while (num[j] == num[j + 1]){ ++ j;}
           ++ j;
         }else{
           ret.push_back({num[i], num[j], num[k]});
           while(num[j] == num[j+1]){ ++j; }
           ++ j;
         }
       }
       while (num[i] == num[i + 1]) ++i;
       ++ i;
     }
     return ret;
   }
};

#endif // LEETLINTCODE_LINTCODE_THREESUM_CPP

File: leet_lint_code/leet_code/java/ThreeSum.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 15.
// Date: 2019.8.26

import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;

public class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);

        int i = 0;

        while (i < nums.length - 2) {
            if (nums[i] > 0) {
                return ans;
            }

            int left = i + 1;
            int right = nums.length - 1;
            while (left < right) {
                List<Integer> list = new ArrayList<>();
                int sum = nums[i] + nums[left] + nums[right];

                if (sum == 0) {
                    list.add(nums[i]);
                    list.add(nums[left]);
                    list.add(nums[right]);
                    ans.add(list);
                    while (right >= 0 && nums[right - 1] == nums[right]) -- right;
                    -- right;
                    while (left < nums.length - 1&& nums[left] == nums[left + 1]) ++ left;
                    ++ left;
                } else if (sum < 0){
                    while (left < nums.length - 1&& nums[left] == nums[left + 1]) ++ left;
                    ++ left;

                } else {
                    while (right >= 0 && nums[right - 1] == nums[right]) -- right;
                    -- right;
                }
            }
            ++ i;
        }

        return ans;
    }
}
